src/blockstore.js

```diff
@@ -55,142 +47,117 @@ function createBaseStore (store) {
+    /**
+     * Write a single block to the store.
+     *
+     * @param {Block} block
+     * @returns {Promise<void>}
+     */
+    async put (block) {
       if (!Block.isBlock(block)) {
-        return setImmediate(() => {
-          callback(new Error('invalid block'))
-        })
+        throw new Error('invalid block')
       }
 
       const k = cidToDsKey(block.cid)
-
-      store.has(k, (err, exists) => {
-        if (err) { return callback(err) }
-        if (exists) { return callback() }
-
-        store.put(k, block.data, callback)
-      })
+      const exists = await store.has(k)
+      if (exists) return
+      return store.put(k, block.data)
     },
+
     /**
      * Like put, but for more.
      *
      * @param {Array<Block>} blocks
-     * @param {function(Error)} callback
-     * @returns {void}
+     * @returns {Promise<void>}
      */
-    putMany (blocks, callback) {
+    async putMany (blocks) {
       const keys = blocks.map((b) => ({
         key: cidToDsKey(b.cid),
         block: b
       }))
 
       const batch = store.batch()
-      reject(keys, (k, cb) => store.has(k.key, cb), (err, newKeys) => {
-        if (err) {
-          return callback(err)
-        }
 
-        newKeys.forEach((k) => {
+      await Promise.all(
+        keys.map(async k => {
+          if (await store.has(k.key)) {
+            return
+          }
+
           batch.put(k.key, k.block.data)
         })
+      )
 
-        batch.commit(callback)
-      })
+      return batch.commit()
     },
     /**
      * Does the store contain block with this cid?
      *
      * @param {CID} cid
-     * @param {function(Error, bool)} callback
-     * @returns {void}
+     * @returns {Promise<bool>}
      */
-    has (cid, callback) {
+    async has (cid) {
       if (!CID.isCID(cid)) {
-        return setImmediate(() => {
-          callback(new Error('Not a valid cid'))
-        })
+        throw new Error('Not a valid cid')
       }
 
-      store.has(cidToDsKey(cid), (err, exists) => {
-        if (err) return callback(err)
-        if (exists) return callback(null, true)
-
-        // If not found, we try with the other CID version.
-        const otherCid = cidToOtherVersion(cid)
-        if (!otherCid) return callback(null, false)
-
-        store.has(cidToDsKey(otherCid), callback)
-      })
+      const exists = await store.has(cidToDsKey(cid))
+      if (exists) return exists
+      const otherCid = cidToOtherVersion(cid)
+      if (!otherCid) return false
+      return store.has(cidToDsKey(otherCid))
     },
     /**
      * Delete a block from the store
      *
      * @param {CID} cid
-     * @param {function(Error)} callback
-     * @returns {void}
+     * @returns {Promise<void>}
      */
-    delete (cid, callback) {
+    delete (cid) {
       if (!CID.isCID(cid)) {
-        return setImmediate(() => {
-          callback(new Error('Not a valid cid'))
-        })
+        throw new Error('Not a valid cid')
```

Comment:
Needs to be `async` or users can't `delete(cid).catch(console.log)` this error.

There's quite a few places where this happens - we should decide if allowing library consumers to use promise syntax is important.

cc @ipfs/wg-js-core

Reply:
Marking things async creates a Promise.  Best case scenario there's no actual async in the function, but the VM will still put the code after the promise onto the microtask queue which creates a small amount of latency.  On hot codepaths this adds up as the Nearform performance analysis of our codebase showed.  Our linting rules enforce [`require-await`](https://eslint.org/docs/rules/require-await) for this reason (and would cause the suggested change to fail linting).

I take the point about promise syntax - it's not great to have an API where some functions are async and other aren't, but we should not create artificial asynchronicity for the reason outlined above.

Maybe I'm wrong but I'd consider this module an internal API, quite different to `js-ipfs` itself so the inconsistency bothers me, but not to the point that we should make the code less efficient.

Reply:
I'm curious to know if `await syncFn()` has the same effect.

We're always going to call functions that may or may not be async with `await` right? - because we just don't know if the function will return a promise or not.

So, if `await` also forces async behaviour then is there a benefit to not using `async`?

Reply:
> If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#Description

Reply:
Exactly why you shouldn't just `await` on everything 😁

FWIW the future-tech version of IPLD figures out if the return value of a function is `then`able and resolves it if so (or at least it did in one iteration).  Making everything async by default prevents these sorts of optimisations.

Anyway I've updated the PR, everything in the docs that claims to be async is now marked async.